### ISSUE TITLE
Instructions on verifying the integrity of Vagrant images on the CentOS blog are misleading

### ISSUE THREAD

**Issue by tjanez** [CentOS/sig-cloud-instance-build#118]
(I'm reporting this here since I don't know if it is possible to adequately format such a long WordPress comment.)

The (currently latest) [blog post on the release of CentOS Vagrant Images version 1801.02](https://seven.centos.org/2018/01/updated-centos-vagrant-images-available-v1801-01/) instructs the user to first download and verify the SHA256 checksums of the images:

```bash
$ curl http://cloud.centos.org/centos/7/vagrant/x86_64/images/sha256sum.txt.asc -o sha256sum.txt.asc
$ gpg --verify sha256sum.txt.asc
```
If the check passed, the user is instructed to use the appropriate checksum to check the downloaded Vagrant image.

### Issue

As Vagrant's `vagrant box add --help` says:

```
Usage: vagrant box add [options] <name, url, or path>

... trimmed ...

The options below only apply if you're adding a box file directly,
and not using a Vagrant server or a box structured like 'user/box':

        --checksum CHECKSUM          Checksum for the box
        --checksum-type TYPE         Checksum type (md5, sha1, sha256)
        --name BOX                   Name of the box
    -h, --help                       Print this help
```

an image's checksum is only checked if a box is added directly.

The command given in the blog post would not report a checksum error, even if it is clearly wrong (e.g. if we specify the checksum to be `deadbeef`):

```bash
$ vagrant box add --checksum-type sha256 --checksum deadbeef --provider libvirt --box-version 1801.02 centos/7
==> box: Loading metadata for box 'centos/7'
    box: URL: https://vagrantcloud.com/centos/7
==> box: Adding box 'centos/7' (v1801.02) for provider: libvirt
    box: Downloading: https://vagrantcloud.com/centos/boxes/7/versions/1801.02/providers/libvirt.box
==> box: Successfully added box 'centos/7' (v1801.02) for 'libvirt'!
```

### Expected results

We can add a Vagrant box manually and check if the checksum is verified:

```
$ vagrant box add --checksum-type sha256 --checksum deadbeef --provider libvirt --name centos/7 https://cloud.centos.org/centos/7/vagrant/x86_64/images/CentOS-7-x86_64-Vagrant-1801_02.Libvirt.box
==> box: Box file was not detected as metadata. Adding it directly...
==> box: Adding box 'centos/7' (v0) for provider: libvirt
    box: Downloading: https://cloud.centos.org/centos/7/vagrant/x86_64/images/CentOS-7-x86_64-Vagrant-1801_02.Libvirt.box
    box: Calculating and comparing box checksum...
The checksum of the downloaded box did not match the expected
value. Please verify that you have the proper URL setup and that
you're downloading the proper file.

Expected: deadbeef
Received: 4440a10744855ec2819d726074958ad6cff56bb5a616f6a45b0a42d602aa1154
```

Adding the box manually actually verified the passed checksum (`deadbeef`) and didn't add the box to Vagrant.

### Solution

Adding a box directly as it was described above does not allow one to specify a box's version by passing the `--box-name` parameter.

A simple solution is to construct a [box's JSON metadata file](https://www.vagrantup.com/docs/boxes/format.html#box-metadata) with the following command:

```
$ cat << EOF > centos7.json
> {
>   "name": "centos/7",
>   "versions": [
>     {
>       "version": "1801.02",
>       "providers": [
>         {
>           "name": "libvirt",
>           "url": "https://cloud.centos.org/centos/7/vagrant/x86_64/images/CentOS-7-x86_64-Vagrant-1801_02.Libvirt.box",
>           "checksum_type": "sha256",
>           "checksum": "4440a10744855ec2819d726074958ad6cff56bb5a616f6a45b0a42d602aa1154"
>         }
>       ]
>     }
>   ]
> }
> EOF
```

This metadata allows one to specify a box's version and also pass the checksum type and value to be veriified.

And it works as expected:
```
$ vagrant box add centos7.json
==> box: Loading metadata for box 'centos7.json'
    box: URL: file:///home/tadej/centos7.json
==> box: Adding box 'centos/7' (v1801.02) for provider: libvirt
    box: Downloading: https://cloud.centos.org/centos/7/vagrant/x86_64/images/CentOS-7-x86_64-Vagrant-1801_02.Libvirt.box
    box: Calculating and comparing box checksum...
==> box: Successfully added box 'centos/7' (v1801.02) for 'libvirt'!
```

**Comment by tjanez** [comment]
@lpancescu, I've seen you've prepared the [latest blog post for the 1803.01 Vagrant images](https://blog.centos.org/2018/04/updated-centos-vagrant-images-available-v1803-01/#comments).

Can you please look at this issue and correct the misleading instructions on verifying the integrity of Vagrant images?

**Comment by kbsingh** [comment]
thanks for the feedback, I'll ping @lpancescu around this for the 1804 announce

**Comment by lpancescu** [comment]
@tjanez I know about the Vagrant bug (well, not inserting the checksums is actually a missing feature from ~Atlas~Vagrant Cloud). I added support for automatically building complete JSON files for all our images to [endymion][] quite a while ago, but people want to just run `vagrant init centos/7`.

One of the most enticing features of Vagrant is to be able to provide a Vagrantfile to less technical users so they can simply run `vagrant up`; a separate JSON file and additional commands for them to run would be counterproductive. It's possible to do everything via the Vagrantfile, I think I'll document that in the next release notes (it's what I've been using myself for the last several weeks).

[endymion]: https://github.com/lpancescu/endymion

**Comment by tjanez** [comment]
> @tjanez I know about the Vagrant bug (well, not inserting the checksums is actually a missing feature from ~~Atlas~~Vagrant Cloud). I added support for automatically building complete JSON files for all our images to endymion quite a while ago, but people want to just run `vagrant init centos/7`.

@lpancescu, thanks for the pointer, I didn't know about endymion.

> One of the most enticing features of Vagrant is to be able to provide a Vagrantfile to less technical users so they can simply run `vagrant up`; a separate JSON file and additional commands for them to run would be counterproductive. It's possible to do everything via the Vagrantfile, I think I'll document that in the next release notes (it's what I've been using myself for the last several weeks).

Agreed, my proposed solution is not easy and convenient. I think your `Vagrantfile` approach is a good alternative for the blog post, because it will cover 90% of people's use cases.
And I see you already implemented it for the [1804.02 Vagrant images blog post](https://blog.centos.org/2018/05/updated-centos-vagrant-images-available-v1804-02/). Thanks!

In my case, I have many Vagrant environments, thus, it would be very inconvenient to have these checks duplicated in all different `Vagrantfile`s.
